python/src/scipp/compat/dict.py

```diff
@@ -0,0 +1,131 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+# @author Neil Vaytet
+
+from .. import _utils as su
+from .._scipp import core as sc
+
+
+def to_dict(scipp_obj):
+    """
+    Convert a scipp object (Variable, DataArray or Dataset) to a python dict.
+
+    :param scipp_obj: A Variable, DataArray or Dataset to be converted to a
+                      python dict.
+    :type scipp_obj: Union[Variable, DataArray, Dataset]
+    :return: A dict containing all the information necessary to fully define
+             the supplied scipp object.
+    :rtype: dict
+    """
+    if su.is_variable(scipp_obj):
+        return _variable_to_dict(scipp_obj)
+    elif su.is_data_array(scipp_obj):
+        return _data_array_to_dict(scipp_obj)
+    elif su.is_dataset(scipp_obj):
+        # TODO: This currently duplicates all coordinates that would otherwise
+        # be at the Dataset level onto the individual DataArrays. We are also
+        # manually duplicating all attributes, since these are not carried when
+        # accessing items of a Dataset.
+        out = {}
+        copy_attrs = len(scipp_obj.attrs.keys()) > 0
+        for name, item in scipp_obj.items():
+            out[name] = _data_array_to_dict(item)
+            if copy_attrs:
+                for key, attr in scipp_obj.attrs.items():
+                    out[name]["attrs"][key] = _variable_to_dict(attr)
+        return out
+
+
+def _variable_to_dict(v):
+    """
+    Convert a scipp Variable to a python dict.
+    """
+    return {
+        "dims": _dims_to_strings(v.dims),
+        "shape": v.shape,
+        "values": v.values,
+        "variances": v.variances,
+        "unit": v.unit,
+        "dtype": v.dtype
+    }
+
+
+def _data_array_to_dict(da):
+    """
+    Convert a scipp DataArray to a python dict.
+    """
+    out = {"coords": {}, "masks": {}, "attrs": {}}
+    for key in out.keys():
+        for name, item in getattr(da, key).items():
+            out[key][str(name)] = _variable_to_dict(item)
+    out["data"] = _variable_to_dict(da.data)
```

Comment:
I'm not a fan of having the same information twice, either via `data.value` or `values` for example. I would say just keep the `data` key.

Reply:
ok